### PR TITLE
Add functions to collate deepspeed zero 3 checkpoints

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Required for #8691, follow-up from https://github.com/PyTorchLightning/pytorch-lightning/pull/8397

Adds a utility as suggested by @kaushikb11 to collate deepspeed sharded checkpoints. This is nicer than pointing users to a random gist by me :)

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/collate_deepspeed_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env python
+# Modified script from https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/utils/zero_to_fp32.py
```
Comment (Borda) [inline pytorch_lightning/utilities/collate_deepspeed_checkpoint.py:2]: do you think it is enought?

File: tests/utilities/test_deepspeed_collate_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+
+@RunIf(min_gpus=2, deepspeed=True, special=False)
+def test_deepspeed_collate_checkpoint(tmpdir):
+    """
+    Test to ensure that with DeepSpeed Stage 3 we can collate the sharded checkpoints into a single file.
+    """
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir, plugins=[DeepSpeedPlugin(stage=3)], gpus=2, fast_dev_run=True, precision=16
```
Comment (kaushikb11) [inline tests/utilities/test_deepspeed_collate_checkpoint.py:33]: We would need to let go of using `plugins` for passing TrainingType Plugins 😛

File: pytorch_lightning/utilities/collate_deepspeed_checkpoint.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/utilities/collate_deepspeed_checkpoint.py:2]: Very good point, as a gist yes, but in our repo no :) let me update the licence!

File: pytorch_lightning/utilities/collate_deepspeed_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    # State dict keys will include reference to wrapper LightningDeepSpeedModule
+    # Delete `module` prefix before saving.
+    state_dict = {k.partition("module.")[2]: state_dict[k] for k in state_dict.keys()}
+    client_state["state_dict"] = state_dict
+
+    print(f"Saving fp32 state dict to {output_file}")
+    torch.save(client_state, output_file)
+
+
+if __name__ == "__main__":
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/collate_deepspeed_checkpoint.py:88]: I was thinking more of it as a utility function than a helper script.
As we don't keep helper scripts inside Lightning.

File: pytorch_lightning/utilities/collate_deepspeed_checkpoint.py (hunk 1 above)
Comment (SeanNaren) [reply pytorch_lightning/utilities/collate_deepspeed_checkpoint.py:2]: I added the original MS licence! (MIT)

File: pytorch_lightning/utilities/deepspeed.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+# Licensed under the MIT license.
+# Modified script from https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/utils/zero_to_fp32.py
+#
+# This script extracts fp32 consolidated weights from a zero 2 and 3 DeepSpeed checkpoints. It gets
+# copied into the top level checkpoint dir, so the user can easily do the conversion at any point in
+# the future. Once extracted, the weights don't require DeepSpeed and can be used in any
+# application. Additionally the script has been modified to ensure we keep the lightning state inside the state dict
+# for being able to run Model.load_from_checkpoint('...').
+#
+# example usage within the lightning checkpoint directory where 'latest' is found:
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/deepspeed.py:12]: Need to update this

File: pytorch_lightning/utilities/deepspeed.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            raise ValueError(f"Unable to find 'latest' file at {latest_path}")
+
+    directory = os.path.join(checkpoint_dir, tag)
+
+    if not os.path.isdir(directory):
+        raise FileNotFoundError(f"Directory '{ds_checkpoint_dir}' doesn't exist")
+    return directory
+
+
+def convert_zero_checkpoint_to_fp32_state_dict(checkpoint_dir, output_file, tag=None):
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/deepspeed.py:47]: Please add type hints

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/utilities/deepspeed.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+from pytorch_lightning.utilities import _DEEPSPEED_AVAILABLE
+
+if _DEEPSPEED_AVAILABLE:
+    from deepspeed.utils.zero_to_fp32 import (
+        get_fp32_state_dict_from_zero_checkpoint,
+        get_model_state_file,
+        get_optim_files,
+    )
+
+device = torch.device("cpu")
```
Comment (Borda) [inline pytorch_lightning/utilities/deepspeed.py:34]:
```suggestion
CPU_DEVICE = torch.device("cpu")
```

**Review by Borda — COMMENTED** [review]
how much is it our original code?

**Review by tchaton — APPROVED** [review]

**Comment by SeanNaren** [comment]
> how much is it our original code?

I found out you can bundle MIT/apache together under the apache licence, so i've done that and included lightning in the licence :)

**Review by Borda — APPROVED** [review]
